lib/req/utils.ex

```diff
@@ -147,22 +148,25 @@ defmodule Req.Utils do
     url = normalize_url(url)
     service = to_string(service)
 
-    canonical_query_string =
-      URI.encode_query([
-        {"X-Amz-Algorithm", "AWS4-HMAC-SHA256"},
-        {"X-Amz-Credential", "#{access_key_id}/#{date_string}/#{region}/#{service}/aws4_request"},
-        {"X-Amz-Date", datetime_string},
-        {"X-Amz-Expires", expires},
-        {"X-Amz-SignedHeaders", "host"}
-      ])
+    canonical_headers =
+      headers
+      |> canonical_host_header(url)
+      |> format_canonical_headers()
 
-    canonical_headers = canonical_host_header([], url)
+    signed_headers = Enum.map_join(canonical_headers, ";", &elem(&1, 0))
 
-    signed_headers =
-      Enum.map_intersperse(
-        Enum.sort(canonical_headers),
-        ";",
-        &String.downcase(elem(&1, 0), :ascii)
+    canonical_query_string =
+      URI.encode_query(
+        [
+          {"X-Amz-Algorithm", "AWS4-HMAC-SHA256"},
+          {"X-Amz-Credential",
+           "#{access_key_id}/#{date_string}/#{region}/#{service}/aws4_request"},
+          {"X-Amz-Date", datetime_string},
+          {"X-Amz-Expires", expires},
+          {"X-Amz-SignedHeaders", signed_headers}
+        ],
+        # Ensure spaces are encoded as %20 not +
```

Comment:
While this should not be a problem with the existing implementation (because arbitrary query params are not supported), this felt like a potential bug waiting to happen, so I updated to use the correct whitespace encoding.